src/v/kafka/data/cloud_topic_partition.cc

```diff
@@ -0,0 +1,599 @@
+};
+
+static ss::future<> bg_upload_and_replicate(
+  ss::shared_ptr<experimental::cloud_topics::data_plane_api> api,
+  ss::lw_shared_ptr<cluster::partition> partition,
+  model::record_batch_header header,
+  ss::lw_shared_ptr<upload_and_replicate_stages> op) {
+    vassert(api != nullptr, "cloud topics api is not initialized");
+    auto fallback = ss::defer([op] {
+        // This guarantees that the promises are set.
+        // The error code used here does not represent the
+        // actual error.
+        op->request_enqueued.set_value();
+        op->replicate_finished.set_value(raft::errc::timeout);
+    });
+    auto timeout = op->timeout == 0ms ? 1s : op->timeout;
+    auto res = co_await api->write_and_debounce(
+      op->ntp, std::move(op->batches), timeout);
+
+    if (res.has_error()) {
+        vlog(
+          kdlog.debug,
+          "LO object upload has failed: {}",
+          res.error().message());
+        co_return;
+    }
+
+    chunked_vector<model::record_batch_header> headers;
+    headers.push_back(header);
+    auto placeholders = convert_to_placeholders(
+      res.value(), std::move(headers));
+
+    vassert(
+      placeholders.batches.size() == 1,
+      "Expected single batch, got {}",
+      placeholders.batches.size());
+
+    // Replicate
+    auto replicate_stages = partition->replicate_in_stages(
+      op->batch_id, std::move(placeholders.batches.front()), op->opts);
+
+    fallback.cancel();
+
+    // Forward future result to the 'op'. The expectation is that at this point
+    // the target promises (inside 'op') are used to generate futures and these
+    // futures are awaited.
+    replicate_stages.request_enqueued.forward_to(
+      std::move(op->request_enqueued));
+
+    auto replicate_fut = std::move(replicate_stages.replicate_finished)
+                           .then(
+                             [](result<cluster::kafka_result> res)
+                               -> result<raft::replicate_result> {
+                                 if (res.has_error()) {
+                                     return res.error();
+                                 }
+                                 return raft::replicate_result{
+                                   model::offset(res.value().last_offset())};
+                             });
+
+    replicate_fut.forward_to(std::move(op->replicate_finished));
+}
+
+ss::future<result<model::offset>> cloud_topic_partition::replicate(
+  chunked_vector<model::record_batch> batches, raft::replicate_options opts) {
+    using ret_t = result<model::offset>;
+    chunked_vector<model::record_batch_header> headers;
+    headers.reserve(batches.size());
+    for (const auto& batch : batches) {
+        headers.push_back(batch.header());
+    }
+    // TODO: use a config
+    auto default_timeout = 1s;
+
+    // Dataplane.
+    auto res = co_await _ct_api->write_and_debounce(
+      ntp(), std::move(batches), opts.timeout.value_or(default_timeout));
```

Comment:
where does the `opts.timeout` value come from? Should we be respecting that for cloud topics?